### PR TITLE
Report warning for implicit conversion of method group to object

### LINKED ISSUE #55923: Report warning for implicit conversion of method group to `object`
Report a warning for an implicit conversion of a method group to `object` when compiling with language version 10 or above, to catch cases where the intent was to invoke the method rather than implicitly create a delegate instance.

```csharp
class Program
{
    static object GetValue() => 0;

    static void Main()
    {
        object value = GetValue; // warning: converting method group to 'object'
    }
}
```

With language version 9, the assignment above is an error:
```
error CS0428: Cannot convert method group 'GetValue' to non-delegate type 'object'.
    Did you intend to invoke the method?
```
With language version 10, the assignment is an implicit conversion of a delegate.

See [LDM-2021-08-23](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-08-23.md#conversions-from-method-group-to-object).

Relates to test plan https://github.com/dotnet/roslyn/issues/52192

### PR DESCRIPTION
**Pull request by cston**
See [LDM-2021-08-23](https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-08-23.md#conversions-from-method-group-to-object).

Fixes #55923

Relates to test plan https://github.com/dotnet/roslyn/issues/52192

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -1356,6 +1356,12 @@
   <data name="ERR_MethGrpToNonDel" xml:space="preserve">
     <value>Cannot convert method group '{0}' to non-delegate type '{1}'. Did you intend to invoke the method?</value>
   </data>
+  <data name="WRN_MethGrpToNonDel" xml:space="preserve">
+    <value>Converting method group '{0}' to non-delegate type '{1}'. Did you intend to invoke the method?</value>
+  </data>
+  <data name="WRN_MethGrpToNonDel_Title" xml:space="preserve">
+    <value>Converting method group to non-delegate type</value>
+  </data>
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:1364]: nit: Doesn't feel like a complete statement like most other warning titles.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -569,6 +569,7 @@ private BoundExpression CreateFunctionTypeConversion(SyntaxNode syntax, BoundExp
             }
 
             conversion = Conversions.ClassifyConversionFromExpression(source, delegateType, ref useSiteInfo);
+            bool warnOnMethodGroupConversion = source.Kind == BoundKind.MethodGroup && conversion.Exists && destination.SpecialType == SpecialType.System_Object;
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:572]: This doesn't check whether the conversion is implicit, is that intended?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:1364]: This is the normal style our warning titles have. Consider `WRN_AmbiguousXMLReference_Title` just above this, for example.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:572]: I would expect us to be checking `isCast`. If we're not, there's probably some test case that's missing.

**Comment by 333fred** [comment]
Done review pass (commit 2).

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 3]
```diff
@@ -1356,6 +1356,12 @@
   <data name="ERR_MethGrpToNonDel" xml:space="preserve">
     <value>Cannot convert method group '{0}' to non-delegate type '{1}'. Did you intend to invoke the method?</value>
   </data>
+  <data name="WRN_MethGrpToNonDel" xml:space="preserve">
+    <value>Converting method group '{0}' to non-delegate type '{1}'. Did you intend to invoke the method?</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:1360]: >'{0}' [](http://example.com/codeflow?start=35&length=5)

I would remove the '{0}' placeholder completely.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:1364]: @333fred Looks like both styles are used.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:572]: > This doesn't check whether the conversion is implicit

Good catch, thanks.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:1360]: I went for consistency with `ERR_MethGrpToNonDel` above.

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:1364]: Will leave as is.

**Review by Youssef1313 — APPROVED** [review]

**Review by 333fred — APPROVED** [review]
LGTM (commit 3)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

File: src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs [hunk 4]
```diff
@@ -1405,6 +1414,154 @@ static class B
...
+        object y = GetValue;
+        y = GetValue;
+        y = (object)GetValue;
+#pragma warning restore 8974
+    }
+}";
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular9);
+            comp.VerifyDiagnostics(
+                // (6,20): error CS0428: Cannot convert method group 'GetValue' to non-delegate type 'object'. Did you intend to invoke the method?
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/DelegateTypeTests.cs:1440]: @cston Should this suggest updating to C# 10 instead?

**Pull request merged**